internal/transport/grpcframer/framer.go

```diff
@@ -0,0 +1,267 @@
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+
+	// Type is a byte that represents the Frame Type. The HTTP/2 spec
+	// defines 10 standard types but extension frames may be written.
+	Type FrameType
+
+	// Flags is a byte representing the flags set for the specific frame
+	// type.
+	Flags Flags
+
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents any kind of HTTP/2 Frame, which can be casted into
+// its corresponding type using the Header method provided.
+type Frame interface {
+	// Header returns this frame's Header.
+	Header() FrameHeader
+}
+
+type DataFrame struct {
+	hdr  FrameHeader
+	pool *sync.Pool
+	Data []byte
+}
+
+func (f *DataFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.Data == nil {
+		return
+	}
+
+	f.pool.Put(&f.Data)
+	f.Data = nil
+}
+
+type HeadersFrame struct {
+	hdr      FrameHeader
+	pool     *sync.Pool
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.HdrBlock == nil {
+		return
+	}
+
+	f.pool.Put(&f.HdrBlock)
+	f.HdrBlock = nil
+}
+
+type RSTStreamFrame struct {
+	hdr  FrameHeader
+	Code ErrorCode
+}
+
+func (f *RSTStreamFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+type SettingsFrame struct {
+	hdr      FrameHeader
+	pool     *sync.Pool
+	settings []byte
+}
+
+func (f *SettingsFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *SettingsFrame) Free() {
+	if f.settings == nil {
+		return
+	}
+
+	f.pool.Put(&f.settings)
+	f.settings = nil
+}
+
+type PingFrame struct {
+	hdr  FrameHeader
+	pool *sync.Pool
+	Data []byte
+}
+
+func (f *PingFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.Data == nil {
+		return
+	}
+
+	f.pool.Put(&f.Data)
+	f.Data = nil
+}
+
+type GoAwayFrame struct {
+	hdr          FrameHeader
+	pool         *sync.Pool
+	LastStreamID uint32
+	Code         ErrorCode
+	DebugData    []byte
+}
+
+func (f *GoAwayFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *GoAwayFrame) Free() {
+	if f.DebugData == nil {
+		return
+	}
+
+	f.pool.Put(&f.DebugData)
+	f.DebugData = nil
+}
+
+type WindowUpdateFrame struct {
+	hdr FrameHeader
+	Inc uint32
+}
+
+func (f *WindowUpdateFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+type ContinuationFrame struct {
+	hdr      FrameHeader
+	pool     *sync.Pool
+	HdrBlock []byte
+}
+
+func (f *ContinuationFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *ContinuationFrame) Free() {
+	if f.HdrBlock == nil {
+		return
+	}
+
+	f.pool.Put(&f.HdrBlock)
+	f.HdrBlock = nil
+}
+
+// GRPCFramer represents the methods a framer must implement to be used in gRPC-Go.
+type GRPCFramer interface {
```

Comment:
This name would appear at callsites as `grpcframer.GRPCFramer` which I feel is quite a mouthful. If we go with the approach of moving this functionality to `internal/grpchttp2`, and name this type `Framer`, it would appear as `grpchttp2.Framer` which is a little better I think. Although not as good as `http2.Framer`, but we do want to clearly indicate that this is a different framer, so `grpchttp2.Framer` won't be so bad.

But on the other hand, given that this is going to be an `internal` package and given that we are going to eventually get rid of all our dependency on the `x/net/http2` package, `http2.Framer` doesn't sound bad either.

Thoughts?
@arvindbr8 @dfawley

Reply:
> it would appear as grpchttp2.Framer which is a little better I think. Although not as good as http2.Framer, but we do want to clearly indicate that this is a different framer, so grpchttp2.Framer won't be so bad.

+1 to this.. I think in the interim we need a clear distinction between the two framers, until we have completely migrated over to the new one

Reply:
Thanks.